### PR TITLE
Fix bucket being in REQUESTED state race condition

### PR DESCRIPTION
**Pull request by MichaelBuessemeyer**
### Summary
This PR fixes a rare but old bug: A in-flight requested bucket can be triggered to be destroyed by e.g. hard reloading its layer. Prior to this PR is was possible to request a bunch of buckets, some still being in a loading state and then some async effect reloading the whole layer -> the late arriving bucket data was never discarded (as requested by the reloading), but never re-requested. => A effect or so waiting for the whole batch being pulled waited infinetly.  The infine waiting was fixed by #9795 but now on master the code for just returns 0ed data instead of awaiting a full reload of the requested buckets due to the reload triggered during the data loading. This PR aims to solve this:

**Update (2026-08-18):** Simplified the approach after review feedback (see the comments below — the first approach and its explanation are marked outdated but kept for context, followed by a new comment describing the current approach). Instead of a new bucket state, all failure causes (backend failure, network error, or a request aborted by a concurrent reload) collapse onto the existing `markAsFailed` path (`REQUESTED` → `UNREQUESTED`, `bucketRequestFailed` event) as they already did on master. `Bucket.ensureLoaded()` now treats that event as retryable: it re-requests the bucket and waits again, up to a configurable retry count (default 3), instead of resolving with stale/empty data. A `REQUESTED` bucket is no longer evicted from the cube during a reload (unchanged from master) — only its in-flight request is aborted, and the same bucket object is retried once the abort settles.

How this bug was found: The adjusted volume integration test was flaky exactly due to that reason. There were interleaving `api.data.reloadAllBuckets()` and bucket editing operations. => The reloading only fully refreshing the layer after reaching a saved state -> `api.data.reloadAllBuckets()` is async and was never awaited. In the test always awaiting `api.data.reloadAllBuckets()` would have fixed the problem but the underlying bug (requesting data giving invali 0ed data). The idea here is to keep the "buggy interleaving test schema" and fixing the underlying bug + adding regression tests for this specific case. Because the api call `getDataValue` should never confidently return false data. It should always wait for proper data if there is. If the bucket is a backend failure request, returning null for those buckets is ok and correct imo => As stated and implemented in #9795.

### Steps to test:
-  Run the following on this branch and on master:
  ```shell
   for i in $(seq 1 30); do yarn run vitest run --config vitest_spec.config.ts \
     frontend/javascripts/test/sagas/volumetracing/volumetracing_saga_integration_2.spec.ts || break; done
   ```
- on master: should break at some point
- on this branch:  should succeed happily

- [x] Moreover: Thoroughly use wk monkey testing style: This fix touches bucket loading which is a very crucial part of wk. Volume annotation, layer bucket data loading & saving & refreshing should work! 
- [x] Proofreading should also still work

### Issues:
- fixes reported here: https://scm.slack.com/archives/C5AKLAV0B/p1784531949066909

------
(Please delete unneeded items, merge only when none are left open)
- [x] Added changelog entry (create a `$PR_NUMBER.md` file in `unreleased_changes` or use `./tools/create-changelog-entry.py`)

### CODE REVIEW

**Comment by MichaelBuessemeyer** [comment]
> **⚠️ Update (2026-08-18):** The root-cause analysis below is still accurate. The "Fix" section describes the first approach (a new `DISCARDED` bucket state + identity-tracked pull-queue batches), which was replaced by a simpler fix after review feedback — see the new comment below for the current approach. Kept here for context.

---

## Why the flaky `getDataValue` → `0` is a real code bug (and why the test is correct)

Writing up the root cause of the intermittent failure in `volumetracing_saga_integration_2.spec.ts` (`Brushing/Tracing with already existing backend data`), since the interleaving is subtle. TL;DR: the value the test reads genuinely exists, `getDataValue` **resolves successfully but returns the wrong value (`0`)** because a concurrent bucket reload evicts the bucket mid-read. That's an internal race, not a test problem.

### What the test asks for

```js
// mock "server" now returns 11 for this layer
vi.mocked(...).mockImplementation(createBucketResponseFunction(..., /* fillValue */ 11, 0));

await api.data.reloadAllBuckets();                               // discard stale cache
expect(await api.data.getDataValue(volumeLayer, [0,0,0])).toBe(11);   // read it back
```

`getDataValue` is a public API whose contract is "return the data value at this position". `11` is genuinely available, so a correct implementation must return `11` deterministically. The test uses only public API in the documented way — every async thing that happens afterwards is our own machinery.

### Background

- Reads go through `DataCube.getLoadedBucket`: `getOrCreateBucket(addr)` → `await bucket.ensureLoaded()` → read voxel. `ensureLoaded()` resolves on `bucketLoaded` / `bucketMissing` / `bucketRequestFailed` (the last one added in #9795 — before that it hung).
- `reloadAllBuckets()` is **async** and defers its eviction past an `await`:
  ```js
  async reloadAllBuckets() {
    if (hasVolumeTracings) await Model.ensureSavedState();   // loops with await sleep(...) while a save is pending
    Object.values(this.model.dataLayers).forEach(l => l.cube.removeAllBuckets());   // the actual eviction
  }
  ```
  So the eviction runs *later*, after saving settles — and it reads `this.model.dataLayers` **at eviction time**, not at call time.

### What actually happens (measured)

Instrumenting every `reloadAllBuckets` call in a failing run shows the eviction steps completing **late and out of order**:

```
CALL#4 ENTER   (earlier "undo/redo" test, line ~280 — called WITHOUT await)
CALL#5 ENTER
CALL#6 ENTER
   "Ignoring request to save because an operation is ongoing. Waiting..."   ← saves queued
... many other reloads enter & evict ...
CALL#9 ENTER   (the Brushing test's own reload, line 382)
CALL#9 DID EVICT
CALL#4 DID EVICT   ← CALL#4 finally evicts HERE, during the Brushing test
```

Sequence of the failure:

1. An **earlier** test fires `api.data.reloadAllBuckets()` **without awaiting** (the undo/redo test around lines 280–293). That reload parks inside `await Model.ensureSavedState()` because saves are still in flight.
2. Its eviction is deferred so long that it outlives the `afterEach`/`beforeEach` (which re-`fetch`es the Model and builds fresh layers) and fires **during a later test** (`Brushing`).
3. That later test — correctly — has already `await`ed its own reload and called `getDataValue([0,0,0])`, which created a fresh bucket, put it in `REQUESTED`, and is `await`ing `ensureLoaded()`.
4. The leaked eviction runs `forEach(this.model.dataLayers…)`, which now points at the **current** test's layers, aborts the in-flight request and **destroys the bucket the read is waiting on**.
5. Destroying it emits `bucketRequestFailed` → `ensureLoaded()` resolves → `getLoadedBucket` hands back the detached, empty bucket → `getDataValue` returns `0`.

### Why this is a code bug, not a test bug

- **A public API resolving with wrong data is a bug by definition.** `getDataValue` doesn't throw or reject — it *successfully returns* `0` when the answer is `11`. Nothing downstream can guard against "sometimes silently wrong".
- **The failing test is correct and is a victim.** It awaits its reload, then reads. It's undermined by another test's async work leaking across a test boundary — which is possible only because `reloadAllBuckets` reads `this.model.dataLayers` *after* its `await`, so a reload requested for one context can evict a different one.
- **The trigger is a legitimate real-world pattern.** Calling `reloadAllBuckets()` without awaiting is not a test-only quirk — production does it too (e.g. `tracing_view.tsx` WebGL-context-restored handler, UI reload actions, sagas). If a fire-and-forget reload can silently corrupt an unrelated concurrent read, that can bite real users.

The correct invariant is: **a read must reflect real data regardless of any concurrent reload.** A reload means "the cached copy is stale — get a fresh one", so a read caught mid-reload should end up with the fresh value, never `0`.

### The fix (three cooperating parts)

1. **Deterministic eviction** — `DataCube.removeBucketsIf`: synchronously fail + evict the just-aborted `REQUESTED` (non-dirty) buckets and wake their waiters, instead of leaving them half-dead for an async abort handler to resolve later.
2. **Identity-guarded requests** — `PullQueue.pullBatch`: capture the exact bucket *objects* a batch was launched for and only mutate those. A stale/aborted request can no longer write data into, or fail, a freshly-recreated bucket that now occupies the same address.
3. **Self-healing read** — `DataCube.getLoadedBucket`: after `ensureLoaded()`, if the awaited bucket is no longer the one registered at its address (a concurrent reload evicted it), re-fetch the fresh bucket and wait again. A genuine load failure keeps the same object in the cache, so this never loops on real errors.

Together: a concurrent reload evicts the read's bucket → the read notices and re-fetches → the stale request can't interfere with the new bucket → the new bucket loads `11` → `getDataValue` returns `11`, every time.

### Verification

- Before the fix the failure reproduced within **3–8 runs**; after the fix the full spec ran **59 consecutive times with zero failures**.
- Added regression tests, each verified to fail if its corresponding fix is removed:
  - `test/model/binary/cube.spec.ts` — reload evicts a `REQUESTED` bucket; dirty buckets are kept; `getLoadedBucket` retries when a concurrent reload evicts the awaited bucket.
  - `test/model/binary/pullqueue.spec.ts` — a superseded request must not clobber a bucket recreated at the same address.
- `yarn typecheck`, `yarn check-frontend`, `yarn fix-frontend` all clean.

To reproduce locally, run the **whole spec file** in a loop (the race relies on an earlier test's deferred reload, so filtering to a single test with `-t` never triggers it):

```bash
for i in $(seq 1 30); do yarn run vitest run --config vitest_spec.config.ts \
  frontend/javascripts/test/sagas/volumetracing/volumetracing_saga_integration_2.spec.ts || break; done
```

**Comment by MichaelBuessemeyer** [comment]
### How long has this bug existed?

Short version: the **underlying race is old (~4–5 years) but was mostly dormant**, and the **specific "reads 0 instead of the real value" symptom is only ~1 week old**.

The bug is a read (`getDataValue` → `getLoadedBucket` → `ensureLoaded`) colliding with a bucket reload that aborts/evicts the very bucket the read is waiting on. It needs two ingredients to exist:

**The latent race — ~4–5 years**
- `abortRequests` — "Abort bucket pulls if a layer is reloaded" (#4940, **Nov 2020**) — a reload cancels in-flight bucket requests.
- `ensureLoaded()` as the wait-for-data mechanism (#5728 / #5733, **Oct 2021**).

Once both existed (~late 2020 / late 2021), a read overlapping a reload could have its bucket yanked away. So the hazard has been latent for roughly **4–5 years**.

**The symptom you actually see in CI — ~1 week**
For almost all of that time this race could **not** produce a silently-wrong value. Before #9795 (**2026-07-17**), `ensureLoaded()` did not resolve on a failed/aborted bucket, so this exact collision would make the read **hang until timeout**, not return empty data. #9795 made `ensureLoaded()` resolve on `bucketRequestFailed`, which flipped the failure mode from "hangs" to "silently returns 0" — which is why the flaky wrong-value assertion only started appearing about a week ago.

**Summary**

| | when | age |
|---|---|---|
| Latent race (abort-on-reload + `ensureLoaded`) | ~2020–2021 | ~4–5 years |
| This "returns 0" symptom (via #9795) | 2026-07-17 | ~1 week |

Caveat (kept deliberately rough): the older it gets the fuzzier — reload/eviction internals were refactored several times since (e.g. `removeBucketsIf` arrived with Live Collab in 2025), and how often it could actually fire depends on timing that changed over the years. Directionally though: the seed is old and mostly dormant, and #9795 recently turned it into a wrong value instead of a hang.

**Comment by MichaelBuessemeyer** [comment]
> **⚠️ Outdated (2026-08-18):** Refers to the first implementation (the `DISCARDED`-state approach), which was replaced by a simpler fix — see the new comment below. Kept here for context.

---

@philippotto sorry for this long PR, but for me it was quite hard to understand the actual problem. Please see my comment in the description where I tried to describe whats going on. Moreover, I let the AI do 2 posts on the problem with should further help to understand the problem.

I iterated in the fix to make it more understandable and I am much more happy with the current implementation compared to the previous one.

Please before testing & looking at the introduced testing code, take a look at the concept of the implementation and tell me your opinion. I wanted for wait for this before going over the ai generated regression tests and testing proofreading

**Comment by MichaelBuessemeyer** [comment]
### Updated approach (2026-08-18)

Simplified the fix after review feedback. Dropped the new `DISCARDED` bucket state and the identity-tracked pull-queue batches (see the outdated notes above). Current mechanism:

- `removeBucketsIf`/`mayBeGarbageCollected` are unchanged from master: a `REQUESTED` bucket is never evicted by a reload, only its in-flight request is aborted via `pullQueue.abortRequests()`. The same bucket object stays attached at its address.
- Once the aborted request rejects, `PullQueue.pullBatch` marks that (still-attached) bucket as failed as before (`markAsFailed()` → `UNREQUESTED` + `bucketRequestFailed`). This already covers backend failures, network errors and aborted requests alike.
- `Bucket.ensureLoaded()` now treats `bucketRequestFailed` as retryable instead of just resolving: it re-requests the bucket and awaits again, up to a configurable retry count (`retriesLeft`, default 3). It throws once retries are exhausted.

This fixes the bug with a much smaller diff, contained mostly in `bucket.ts`, with `data_cube.ts`/`pullqueue.ts`/`texture_bucket_manager.ts` reverted to their pre-PR state. Tests in `cube.spec.ts`/`pullqueue.spec.ts` updated accordingly (retry-then-succeed, throw-after-retries-exhausted, and a regression test reproducing the original reload-vs-read race). Verified with 15/15 clean runs of the integration-spec loop plus the full test suite.

PR description updated to match.

File: frontend/javascripts/viewer/model/bucket_data_handling/bucket.ts [hunk 1]
```diff
@@ -829,7 +834,10 @@ export class DataBucket {
     }
   };
 
-  async ensureLoaded(): Promise<void> {
+  async ensureLoaded(
+    retryCount: number = 0,
+    maxRetries: number = DEFAULT_ENSURE_LOADED_RETRY_COUNT,
```
Comment (philippotto) [inline frontend/javascripts/viewer/model/bucket_data_handling/bucket.ts]: I'd rename retryCount to currentRetryCount. Apart from that, it's a bit weird that retryCount is a parameter which can theoretically be set from the outside. How about you add an inner function (`const ensureLoadedInner = (currentRetryCount) => ...`) which accepts that parameter?

File: frontend/javascripts/viewer/model/bucket_data_handling/bucket.ts [hunk 2]
```diff
@@ -843,13 +851,24 @@ export class DataBucket {
...
-      await new Promise((resolve) => {
-        this.once("bucketLoaded", resolve);
-        this.once("bucketMissing", resolve);
-        // Treat failure like missing to not block callers waiting for ensureLoaded to resolve
-        // as a failure not necessarily auto re-requests the bucket from the backend.
-        this.once("bucketRequestFailed", resolve);
+      const didRequestFail = await new Promise<boolean>((resolve) => {
+        this.once("bucketLoaded", () => resolve(false));
+        this.once("bucketMissing", () => resolve(false));
+        this.once("bucketRequestFailed", () => resolve(true));
```
Comment (philippotto) [inline frontend/javascripts/viewer/model/bucket_data_handling/bucket.ts]: if bucketRequestFailed triggers, its listener here will be removed thanks to `once`. however, the two other listeners will still be intact. they can fire on the next retry. it's not really a problem because the handler will only try to resolve a promise that was already resolved (noop).
still, I feel like it's a bit hacky. feel free to ignore if you think I'm thinking too much into this.

File: frontend/javascripts/viewer/model/bucket_data_handling/bucket.ts [hunk 3]
```diff
@@ -843,13 +851,24 @@ export class DataBucket {
...
+      if (didRequestFail) {
+        // The request failed. This covers a backend failure, a network error and a request
+        // that was aborted (e.g. because a concurrent reload called pullQueue.abortRequests()).
+        // Thus, we should retry now.
+        if (retryCount >= maxRetries) {
+          throw new Error(
+            `Bucket ${this.zoomedAddress.join(",")} could not be loaded after ${maxRetries} retries.`,
+          );
+        }
+        await this.ensureLoaded(retryCount + 1, maxRetries);
```
Comment (philippotto) [inline frontend/javascripts/viewer/model/bucket_data_handling/bucket.ts]: how about exponential backoff?

**Review by philippotto — CHANGES_REQUESTED** [review]
thanks for taking care of this! see my two small suggestions please :)

File: frontend/javascripts/test/sagas/volumetracing/volumetracing_saga_integration_2.spec.ts [hunk 4]
```diff
@@ -276,22 +276,30 @@ describe("Volume Tracing", () => {
...
-    api.data.reloadAllBuckets();
+    pendingReloads.push(api.data.reloadAllBuckets());
     await assertFloodFilledState();
-    // "Undo", reload all buckets and assert flood-filled state
+    // "Undo", reload all buckets and assert initial state
     await dispatchUndoAsync(Store.dispatch);
-    api.data.reloadAllBuckets();
+    pendingReloads.push(api.data.reloadAllBuckets());
     await assertInitialState();
+    await Promise.all(pendingReloads);
```
Comment (MichaelBuessemeyer) [inline frontend/javascripts/test/sagas/volumetracing/volumetracing_saga_integration_2.spec.ts:302]: The current implementation is awaiting all reloads at the end to still keep the test flaky in case we have a regression to detect it. But there is also an explicit test for this. Should I instead await all `api.data.reloadAllBuckets()` directly?

**Review by MichaelBuessemeyer — COMMENTED** [review]
ok thanks for the feedback. It should now be implemented 👍

File: frontend/javascripts/viewer/model/bucket_data_handling/bucket.ts [hunk 5]
```diff
@@ -838,28 +843,52 @@ export class DataBucket {
...
-        this.once("bucketRequestFailed", resolve);
-      });
-    }
+      if (needsToAwaitBucket) {
+        const didRequestFail = await new Promise<boolean>((resolve) => {
+          const unbindFunctions: (() => void)[] = [];
+          const clearAndResolve = (success: boolean) => {
+            unbindFunctions.forEach((fn) => {
+              fn();
+            });
```
Comment (philippotto) [inline frontend/javascripts/viewer/model/bucket_data_handling/bucket.ts:866]:
```suggestion
            unbindFunctions.forEach(fn => fn());
```

unless typescript complains

File: frontend/javascripts/test/sagas/volumetracing/volumetracing_saga_integration_2.spec.ts (hunk 4 above)
Comment (philippotto) [reply frontend/javascripts/test/sagas/volumetracing/volumetracing_saga_integration_2.spec.ts:302]: > The current implementation is awaiting all reloads at the end to still keep the test flaky 

wasn't the test flaky without these already (current master state)?

> But there is also an explicit test for this

which test is that? the new one in cube.spec.ts?

I think we don't need to put much effort into having good regression tests, because even the old, flaky one led us to the problem. I'm fine with having the new test(s) though.

**Review by philippotto — APPROVED** [review]
great stuff! just see my one suggestion and the reply to your other comment.

File: frontend/javascripts/viewer/model/bucket_data_handling/bucket.ts (hunk 5 above)
Comment (MichaelBuessemeyer) [reply frontend/javascripts/viewer/model/bucket_data_handling/bucket.ts:866]: yeah sorry, biome complained afaik as `fn => fn()` returns the return value of `fn()` even when this is declared to be void. So the wrapping brackets are needed :///// Found this pretty annoying as well :/

File: frontend/javascripts/test/sagas/volumetracing/volumetracing_saga_integration_2.spec.ts (hunk 4 above)
Comment (MichaelBuessemeyer) [reply frontend/javascripts/test/sagas/volumetracing/volumetracing_saga_integration_2.spec.ts:302]: > wasn't the test flaky without these already (current master state)?

Yes, the advantage here is now that we have a better test isolation: The test only finishes after all potential interleaving / stale reloads are done / awaited -> They can't interfere with the next test.

> which test is that? the new one in cube.spec.ts?

Jup, exactly 👍
